srml/im-online/src/lib.rs

```diff
@@ -519,7 +527,11 @@ impl<T: Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
 		where I: Iterator<Item=(&'a T::AccountId, T::AuthorityId)>
 	{
 		// Tell the offchain worker to start making the next session's heartbeats.
-		<GossipAt<T>>::put(<system::Module<T>>::block_number());
+		// Since we consider producing blocks as being online,
+		// the hearbeat is defered a bit to prevent spaming.
+		let block_number = <system::Module<T>>::block_number();
```

Comment:
Why does this prevent spamming?

Reply:
Instead of sending heartbeats at the beginning of the session we will wait a bit and only the validators that didn't have a chance to produce a block so far will send the hearbeats. So roughly we will decrease the amount of "spammy" heartbeat transactions by half - I consider all of those transactions spammy, as they are just part of the inner workings and just "take place" for users transactions / introduce a bit of noise.

Reply:
Ahh makes sense :)